src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java

```diff
@@ -187,57 +202,177 @@ private void dataCmd() {
+  }
+
+  private void sendMailHeaders(List<Map.Entry<String, String>> headers, int i, Handler<Void> endHandler) {
+    if (i < headers.size()) {
+      String entryString = headers.get(i).toString();
+      connection.writeLineWithDrainHandler(entryString, written.getAndAdd(entryString.length()) < 1000,
+        h -> sendMailHeaders(headers, i + 1, endHandler));
+    } else {
+      connection.writeLineWithDrainHandler("", written.get() < 1000, endHandler);
+    }
+  }
+
+  private void sendBodyLineByLine(String[] lines, int i, Handler<Void> endHandler) {
+    if (i < lines.length) {
+      String line = lines[i];
       if (line.startsWith(".")) {
         line = "." + line;
       }
-      final int nextIndexFinal = nextIndex;
-      final boolean mayLog = nextIndex < 1000;
-      if (connection.writeQueueFull()) {
-        connection.writeLineWithDrainHandler(line, mayLog, v -> sendLineByLine(nextIndexFinal, length));
-        // call to our handler will finish the whole message, we just return here
-        return;
-      } else {
-        connection.writeLine(line, mayLog);
-        index = nextIndex;
-      }
+      connection.writeLineWithDrainHandler(line, written.getAndAdd(line.length()) < 1000,
+        h -> sendBodyLineByLine(lines, i + 1, endHandler));
+    } else {
+      endHandler.handle(null);
     }
-    connection.write(".", message -> {
-      log.debug("maildata result: " + message);
-      if (StatusCode.isStatusOk(message)) {
-        resultHandler.handle(Future.succeededFuture(mailResult));
+  }
+
+  private void sendRegularPart(EncodedPart part, Handler<Void> endHandler) {
+    sendMailHeaders(part.headers().entries(), 0, v -> {
+      // send body after headers
+      if (part.body() != null) {
+        // send body string line by line
+        sendBodyLineByLine(part.body().split("\n"), 0, endHandler);
+      } else if (part.bodyStream() != null) {
+        BodyReadStream bodyReadStream = new BodyReadStream(part.bodyStream());
+        bodyReadStream.pipe().endOnComplete(false).to(connection.getSocket(), r -> {
+          if (r.succeeded()) {
+            endHandler.handle(null);
+          } else {
+            log.warn("Failed to transfer stream", r.cause());
+            handleError(r.cause());
+          }
+        });
       } else {
-        log.warn("sending data failed: " + message);
-        handleError("sending data failed: " + message);
+        throw new IllegalStateException("No mail body and stream found");
       }
     });
   }
 
-  /**
-   * create message if it hasn't been already
-   */
-  private void createMailMessage() {
-    if (mailMessage == null) {
-      MailEncoder encoder = new MailEncoder(email, hostname);
-      mailMessage = encoder.encode();
-      mailResult.setMessageID(encoder.getMessageID());
+  private Handler<Void> handlerInContext(Handler<Void> handler) {
+    return vv -> connection.getContext().runOnContext(v -> handler.handle(null));
```

Comment:
i think
```java
return vv -> connection.getContext().runOnContext(handler);
```

Would have the same result and it avoid 1 extra lambda

Reply:
:+1: